backend/db_migrations/000002_auth_system.up.sql

```diff
@@ -0,0 +1,192 @@
+    person_id bigint primary key references app.person(id) on delete cascade,
+    password_hash text not null,
+    created_at timestamp default now() not null,
+    updated_at timestamp default now() not null
+  );
+
+  comment on table app_private.account is 'Table to store person private details';
+
+  create trigger tg_account_set_updated_at
+  before update on app_private.account 
+  for each row execute procedure app.set_updated_at();
+
+  create trigger tg_account_set_created_at before insert
+  on app_private.account 
+  for each row execute procedure app.set_created_at();
+
+  create extension if not exists pgcrypto;
+
+ -- create user --
+
+  create type app.jwt_token as (
+    role text,
+    person_id bigint,
+    exp bigint
+  );
+
+
+  create type app.current_user as (id bigint, role text, email text, full_name text, slug text);
+
+
+create or replace function app.register_person(
+    full_name text,
+    email text,
+    password text
+) returns app.person as $$
+    declare p app.person;
+begin
+    insert into app.person (full_name, email)
+    values (full_name, email)
+    returning * into p;
+    insert into app_private.account(person_id, password_hash)
+    values (p.id, crypt(password, gen_salt('bf')));
+    return p;
+end;
+$$ language plpgsql security definer;
+
+--replacing the function to include app.person.role (from person table) in the jwt_token
+create or replace function app.authenticate(user_email text, password text) returns app.jwt_token as $$
+    declare acct app_private.account;
+    person app.person;
+begin
+    select * into person
+        from app.person p
+        where p.email = user_email;
+    select * into acct
+        from app_private.account a
+        where a.person_id = person.id;
+    -- The salt is stored along with the password_hash. The crypt function
+    -- will read the bytes corresponding to the salt from the password_hash column
+    if acct.password_hash = crypt(password, acct.password_hash) THEN return (
+        person.role::text,
+        acct.person_id,
+        extract(epoch from (now() + interval '7 days'))
+    )::app.jwt_token;
+    end if;
+    return null;
+end;
+$$ language plpgsql security definer;
+ comment on function app.authenticate(text,text) is 'Login method for People. Username and password are authenticated against the account table. On success a JWT is created with claims for Person and role.';
+
+create or replace function app.make_first_user_admin() returns trigger as $$
+    declare person_count int;
+begin
+    select count(1) into person_count
+    from app.person;
+    if person_count = 0 then 
+        new.role = 'app_admin';
+    end if;
+    return new;
+end;
+$$ language plpgsql security definer;
+create trigger tg_person_make_admin 
+before insert on app.person 
+for each row execute procedure app.make_first_user_admin();
+
+create or replace function app.authorize_admin(p_id bigint) returns app.person as $$ 
```

Comment:
Do we need drop for authorize functions as other functions are dropped in down?

Reply:
Done.